### DISCUSSION TITLE
let and var Keywords

### DISCUSSION THREAD

**Discussion by LZL0** [modular/modular#120] [General]
As the documentation states at https://docs.modular.com/mojo/notebooks/HelloMojo.html

`[...] To support this, Mojo supports let and var declarations, which introduce a new scoped runtime value: let is immutable and var is mutable [...].`

I would like to raise some awareness that this might create confusion, arguable the most popular programming language after Python, JavaScript has `const` as immutable with `let` and `var` being mutable. People who would program in both Mojo🔥  and JavaScript/TypeScript in the future would probably appreciate if `let` in Mojo🔥  was mutable instead or remained immutable but renamed to `const`.

**Comment by TZDak** [comment]
I've also got some concerns with how let and var and variable declarations, in general, are currently being handled. If Mojo is meant to be (in the long run) a superset of Python able to run Python stuff within it, the following should be possible:
```
a = 9
print(a)
a = "Hello"
print(a)
```
...but at least currently it's not. The second assignment complains that the string "Hello" cannot be converted to an Int.

So, I would like to propose a modification to your suggestion. If you think I should put it in a new discussion thread, let me know. If you would like to do so, with or without modifications of your own, feel free.

#Mojo Update Proposal:

1: Allow freely typed variables to be created and typed or retyped by means of assignment, without any explicit declaration, as in Python. Perhaps discourage their usage and maybe even make an appropriate compiler warning about it, but for compatibility with Python and ease of early learning, allow it.

Usage example: (Assigns a value of None to the variable named "a" which would be effectively stored as a Python-style dictionary entry, just like Python does. Then assigns the value 42 to the same variable, changing its value and its type.)
```
a = None
a = 42
```

2: Use "var" to create dynamically typed dynamic variables, which require redeclaration to be retyped.
Usage example: (Declares the variable named "b" to be of type Int, and assigns the string value "xmpl" to the variable names "c". The value of either variable can be freely changed afterward, but trying to change the type would raise an exception unless done so in a new declaration.)
```
var b: Int
var c = "xmpl"
b = 9
print(c, b)
var b: StringRef = "Nine"
print(b)
var c = 3.1416
print(c)
```

3: Use "let" to create fixed-typed static variables. Once assigned a type within the given scope, the value of the variable could be changed but the type could not. Trying to change the type of such a variable would cause an exception to be thrown.

4: Use "const" to create strictly typed static variables. Trying to change either the value or the type of a variable once declared with "const" in the given context would throw an exception. (This would take over the functionality currently given by "let".)

**Reply by LZL0** [reply]
I love your 2nd proposal.

**Comment by ksandvik** [comment]
This is similar to variable shadowing. I think it should be fine to be allowed if let or var is used in front of the variable (similar to Rust). But as far as I know the intent with Mojo is to catch and optimize code so having explicit type annotation is fine. As for const, var, let, mut et rest, there's no real standard anyway between the languages so for me Mojo could have its own syntax.

**Comment by ego** [comment]
Perhaps we can keep it simple and save our fingers from typing letters and assume that all staff is immutable (aka `let`) and only mutable variable labels with `var`. This approach is commonly used in the V lang (with keyword `mut`).

**Comment by lattner** [comment]
Re:
```
a = 9
print(a)
a = "Hello"
print(a)
```

I agree we need to decide what the model is.  This /must/ work, at least in a def, for python compatibility.  'def's currently allow implicit declaration, but infer the type from the first assignment.  The above implies that implicitly declared variables in a def should default to having object type (which type erases the concrete type and will allow the above).

I think this is the right/unavoidable thing to do, but I have two concerns:
1) We don't really have the language features in place to implement object correctly (notably need the basics of classes), so I'd like to avoid switching to this model until we can make it work right.

2) This push us to define/create the "type erasure of structs to object" model so that user defined struct types can be used here.  We may or may not want to do this, it isn't clear to me.  There is a lot of precedent in this in the Swift world where Swift classes can be typed erased to "AnyObject" (aka `id` in ObjC) and that type allow dynamic dispatch in various ways.  See eg https://github.com/apple/swift-evolution/blob/main/proposals/0116-id-as-any.md

These are super nuanced issues and I'd like to get more experience with the core language before touching into this.  There is a big difference between bringing up something simple and building it really great.

**Reply by mahmoudajawad** [reply]
I'm all-in for dropping `var` and use `let` for re-defining same variable with different type.

**Reply by sa-** [reply]
>  There is a big difference between bringing up something simple and building it really great

That's a great point, and I'm happy to be patient on the nice-to-haves in favour of building a solid core. 

What I'd like to say to the community is that prioritizing features is what ultimately makes a programming language great (or not), because the features you build in the beginning are the ones that the new features will have to account for and build around in the future. Further, given that the team is small and they can only focus on a small number of things - I'm happy that they are focusing on much needed features like package management, enums, & traits.

**Reply by nmsmith** [reply]
Chris, I made a proposal for how to handle rebinding like this [on the variable shadowing thread](https://github.com/modularml/mojo/issues/5#issuecomment-1537590471). I’d love to hear your thoughts. To summarise, I think it would be very useful to allow a variable’s type to change after a reassignment — even in a `fn`. This has utility in scenarios where you want to “transform” a value in a way that changes its type, e.g. by unwrapping or validating it. This also (both in a `def` and a `fn`) allows types to remain specific rather than falling back on AnyObject etc.

A way to think about this: a variable being “mutable” means that both its value and type can be mutated. This is quite natural in Python. A Python programmer regularly mutates the (dynamic) types of variables. To a Python programmer, to fix a variable’s type based on the first thing assigned to it seems strange.

**Reply by mahmoudajawad** [reply]
> much needed features like package management, enums, & traits.

@sa-, at this point everything is important as any decision could dictate the a long-term language feature that could be proven not best.

**Reply by strangemonad** [reply]
@nmsmith I see where you're coming from but I've generally tried to guide my teams to use this more sparingly to manage complexity as code bases grow. We've even enabled linting modes that prohibit arbitrary reassignment and variable shadowing outside of notebooks.  Rather than a variable being able to be re-assigned to anything, it usually starts life as a well known set of types e.g. 

```python
def polymorphic(a: String | datetime | int):
  if isinstance(a, String):
    ...
  elif isinstance (a, datetime):
    ...
 ...
```

With a proper type system, you can rely on type refinement in these cases.

If you truly think arbitrary re-assignment might be useful, what cases does that tend to be in? Notebooks and exploratory code? Doing correct type inference and type checking in that case is definitely harder (read takes longer) because you have to build up your CFG and check properties on multiple SSA fronts. Would a strict / lax mode as I've tried to sketch out in https://github.com/modularml/mojo/discussions/120#discussioncomment-5952041 be workable for what you're thinking of?

**Reply by nmsmith** [reply]
@strangemonad The idea that I proposed—allowing a variable's type to change upon assignment—does not require any special extensions to the type checker. You would use exactly the same logic that is used for variable shadowing. Literally, just pretend there is a `let` out the front of each assignment. That is the type-checking logic you would use. (I provide more details in the variable shadowing thread.) The only complication is if a user tries to assign a variable different types across the branches of an `if`. There are several ways one could handle that. I enumerate those in [the variable shadowing thread](https://github.com/modularml/mojo/issues/5#issuecomment-1537590471). The simplest approach is just to make it a compile error — just as it is today.

> If you truly think arbitrary re-assignment might be useful, what cases does that tend to be in? Notebooks and exploratory code?

The use-cases are the same as those for variable shadowing. These use-cases were discussed in the variable shadowing thread. One particular example is "unwrapping" something to extract it from the `Optional` type, or relatedly, when parsing something to extract it from a string. In both of these operations, the value (informally) has an "identity" that remains the same after the operation (I'm "extracting" or "processing" something), but the operation changes the value's type.

**Reply by gryznar** [reply]
> Chris, I made a proposal for how to handle rebinding like this [on the variable shadowing thread](https://github.com/modularml/mojo/issues/5#issuecomment-1537590471). I’d love to hear your thoughts. To summarise, I think it would be very useful to allow a variable’s type to change after a reassignment — even in a `fn`. This has utility in scenarios where you want to “transform” a value in a way that changes its type, e.g. by unwrapping or validating it. This also (both in a `def` and a `fn`) allows types to remain specific rather than falling back on AnyObject etc.
> 
> A way to think about this: a variable being “mutable” means that both its value and type can be mutated. This is quite natural in Python. A Python programmer regularly mutates the (dynamic) types of variables. To a Python programmer, to fix a variable’s type based on the first thing assigned to it seems strange.

fn should be strict. No type reassignment. In fn/struct compiler should not do type infering.

**Reply by ivellapillil** [reply]
a = 9
print(a)
a = "Hello"
print(a)

How about inferring the type as a union of Int and String? This would be more tighter than erasing the type...?

**Comment by strangemonad** [comment]
I've also been thinking about this a bunch but mostly from the perspective of gradually adopting mojo into a larger python codebase.

@lattner to your point on type erasure on structs, my initial hunch is that this made a ton of sense for Swift given the history of Obj-c and NSObject / CFObject bridging but would be undesirable here but I'm also still agnostic for the reasons you've pointed out.

My current thinking has led me in another direction that I feel might be less surprising to python devs
- drop the keywords
- if you're in a "mojo context" or "strict context" (handwaves) it's always a let (maybe this is `fn` vs `def` but it's likely more nuanced
- you regain mutability by using some sort of mutable box type e.g. `counter: Var[Int]`
  - There's precedence with other places where python uses generic types like this
  - "oh but we don't want everything to be some sort of ref-cell, that's not capable enough to represent full `var` semantics". I agree. There would be a handful of these mutable box types to choose from (I'm not sure if this opens us up to the complexity of rust's arc / box world?).
  - 90% of the time, I think folks would be well served by a `Var` that has the semantics of java's `Atomic` / clojure's `atom`
- I haven't completely thought this through but I wonder if this might also simplify register-passable fn arguments e.g. `fn foo(a: Reg[Int]): ...`

**Comment by cpsource** [comment]
My vote is for 'const' as a modifier of 'let' and drop the other one. Such a construction is more readable and closer to 'c'.

**Reply by sa-** [reply]
Even though I'm a fan of `let` and `var`, I can see why keeping things mutable and keywordless by default is more pythonic, and introducing a keyword like `const` or `fix` to specify an immutable value would make sense. 

It would certainly be easier to teach programming to my niece (and other non-programmers) with one less keyword.

**Comment by pchemguy** [comment]
I vote for the 'const' keyword too.

**Comment by ksandvik** [comment]
There's also the opposite of everything being immutable and var (or something else, similar to Rust's mut) makes it mutable. I'm open for any version. For me const has a reflection of how it's used in C, compile time lockdown, though. But one less keyword might be good.

**Reply by pchemguy** [reply]
To me, this approach does not make as much sense. IMHO, the mutable variables convention is the "normal/intuitive" default in Python, at least.

**Comment by sa-** [comment]
After thinking about it some more, my personal preferences (for whatever they are worth) are to have keywordless definition of mutable variables (pythonic) and a keyword `fix` for immutable variables. This will prevent confusion from those who expect `const` to be a compile time keyword but it will still convey the right meaning to veteran and newbie programmers alike.

```python
a = 5
fix b = 2
a = b
b = 2  # this would cause an error, since b is immutable
```

**Reply by ksandvik** [reply]
Yes, that conforms to the Pythonian way of variables always being mutable. Then again in Mojo, arguments passed in functions are non-mutable.  This all will be interesting...

**Reply by gryznar** [reply]
I like current Mojo state in which reassigning type is disallowed. I also like explicit declaration of mutability. def may allow mutability, but fn shouldn't without declaration. Python is slow mainly due to the fact that variable can be everything.
Speed for me is great counter and compromising it to allow such dynamic magic is not the right direction for me. 

Struct/fn in my opinion should be designed for static programming, class/def for sort of dynamic.

**Reply by helehex** [reply]
I'd like `fix` if they keep the behavior.. but then again, imo, keywords should avoid intersecting with potential user defined stuff (like if a user want a struct to have a fix() function).
I like when they're an unenglish shortened version of the word that's still descriptive enough. 
ei. struct vs structure, or fn vs function.
Someone reading the code is expected to know the keywords, but has to be told what user defined stuff does

**Comment by gryznar** [comment]
I would suggest to be forced to be explicit about types / mutability etc. in struct / fn if it helps to just squeeze maximum performance and allow compiler to do all possible optimalization. class / def could just behave like in Python.

In global scope, declaring mutability / types etc could allow to do much more optimization, but would not be enforced - for Python compatibility. It may just emit a warning as was suggested sth above

**Comment by gryznar** [comment]
I would like to think about possibilities in strict mode (redeclaring variable with the same name etc.) as "Would it hurt performance?":
- if yes - this may be not reasonable to introduce
- if no - probably we can, if it could help in some way.
Python is considered to be slow due to a lot of performance compromises. Mojo, on the other hand could  compromise ONLY if it is necessary.

**Comment by lattner** [comment]
relatedly: https://github.com/modularml/mojo/blob/main/proposals/lifetimes-keyword-renaming.md#allow-let-in-argument-lists--or-remove-them-entirely-

**Reply by gryznar** [reply]
[> relatedly: https://github.com/modularml/mojo/blob/main/proposals/lifetimes-keyword-renaming.md#allow-let-in-argument-lists--or-remove-them-entirely-](https://github.com/modularml/mojo/discussions/338#discussioncomment-6105185)

Here I shared my concerns about this proposal

**Comment by pchemguy** [comment]
Mutating variable type is a bad practice and most of the time should be unnecessary. There are certain scenarios, where the data source cannot be fully controlled by the program, such as data coming from an SQLite database with non-strict columns or from a JSON file. In those cases, full mutability might in fact be handy. Even though conventionally, Python does support full mutability of variables with dynamic typing, mutating variable's type is a recipe for bugs and a strong obstacle for automatic code analysis tools. In fact, modern Python, while still sticking to the dynamic typing paradigm, increasingly develops the type hinting infrastructure.

Mojo probably should follow the best practices while maintaining compatibility with Python. This means encouraging type hints wherever possible and issuing warnings on non-annotated variables or type mutation.

In fact, you probably do not need a keyword to indicate that a variable should be mutable. Instead, type hint should play such a role, particularly within "non-Python" code blocks, such as `fn`.

Mutating value is a separate matter, and I would support value mutability by default (like in Python). Functional (`fn`) call is a special case. Default immutable values for scalar types with the ability to declare a mutable parameter is a possible approach.

**Comment by cirqol** [comment]
@LZL0 I wouldn't base any language design decisions on Javascript 🤢 not exactly the best role model.

**Reply by LZL0** [reply]
The current usage of the keywords just doesn't make much sense.

**Reply by goyalyashpal** [reply]
> I wouldn't base any language design decisions on Javascript 🤢

i was thinking exactly same. this particular aspect of proposal sounded pretty narrow and selfish - proposing a huge _and_ particular change based on single data point.

**Reply by LZL0** [reply]
In my opinion, changing a keyword in a language that is still being developed shouldn't be a contentious issue. It would be wise to utilize industry-accepted terminology that is widely comprehensible, rather than introducing additional complexity. This approach helps people with both adoption and maintenance of Mojo 🔥 code.

**Reply by goyalyashpal** [reply]
oh now so, one language is the "industry standard" ?
as i already said - based on "single data point" - i would have appreciated if you looked at more breadth of the language landscape - but u didn't. rather u picked what is too convenient to you, and that's why, it's selfish.

**Comment by ksandvik** [comment]
Discussions about keywords will always become very personal and whatever is selected then others will be unhappy. And that is fine. Or, please be polite to each other.

**Reply by gryznar** [reply]
There should be two meta rules in my opinion ;) Clarity and unambiguity. I am not the fan of restricting keywords to absolute minimum. The opposite, if additional keyword will help in controlling low-level staff for example I am on it.

**Comment by gvanrossum** [comment]
On a side note, Python already has something similar to `var x: Type [= value]` -- you can write `x: Type [= value]`. Why not use that? It should work in struct as well as in local and function scopes. There's the issue that the syntax corresponding to `var x = value` has different semantics in Python, but still...

**Reply by nmsmith** [reply]
My current suspicion is that Mojo shouldn't even be using `=` to transfer a value (a struct instance) into a memory location, because in Python, `=` merely defines (or redefines) an _alias_ to an existing memory location, where an object resides. (Technically, yes, `=` transfers an _opaque value_—a reference—into a memory location, but this is not the same thing as transferring a struct/object instance.)

Mojo is actually going to have aliases (reference variables), just as Python does. This is alluded to in [the lifetimes proposal](https://github.com/modularml/mojo/blob/main/proposals/lifetimes-and-provenance.md). Therefore, to be consistent with Python, I'm starting to think that `=` should be used for defining and updating those.

Accordingly, to transfer a value (a struct instance) into a memory location, Mojo might want to introduce a `<-` operator. This would literally mean "put this value into this memory cell". Python has no equivalent concept, because Python objects can (from the user's perspective) never be transferred between memory cells.

**Reply by helehex** [reply]
`a <-- b` explicit deep copy
`b <--- c` explicit deeper copy
`c <--∞ d` very explicit deepest copy

as parameter of function:
```rust
fn deepstuff(d1: deep32, d2: deep32):
 d1.repopulate()
 deepstuff(<--∞ d2, <--∞ d1)
```

**Reply by helehex** [reply]
Some half baked thoughts:

```rust
a: Int <- 5
// "<-" direct assignment operator.
// assigns "5" to "a"
// strictly typed left hand side
// type of right hand side must evaluate at compile time
// calls __copyinit__ for structs and classes
b: Int <- 6
a: Int <- b
// unwraps python reference variables
d: Int <- 4
c = d // "c" is now a pythonic reference to "d"
// structs call __copyinit__ on wrap
b = c
c += 1
a: Int <- b
// "a" is now a direct value of 5
// structs call __copyinit__ on unwrap


a = b
// "=" Indirect assignment operator
// "a" will reference the same data as "b" 
// "pythonic behavior"
// structs get wrapped in a python object and
// overloads copyinit() to behave like a python object
b: Int <- 5
a: Int* = &b
// the same as
b: Int <- 5
a = b
// classes behave like python, and calls copyinit():
b: int = 5
a = b


a <-> b
// direct swap
// sweet sugar
// only calls to __moveinit__

a <=> b
// indirect swap

def function():
 :a: Int
 a <- 5
 // "a" is lexically scoped, with name shadowing, and lazy initialization
 // requires redeclaration to be retyped
 a +<- 3
 // "a" calls Int.iadd(), is incremented by 5
 // as we all know, + is indeed less than -
 :b: Int = 6
 a += b
 // "a" calls Int.iadd(), is incremented by 6
 // "a" is now an Int wrapped in a python object
 // (for good luck)
 a +<- 7
 // a is no longer wrapped in a python object


a = <-b
// assigns a mutable reference to a copy of `b`
// ensures the reference is a copy

a <-- b
// deep assignment operator
// assigns mutable deep copy of `b`
// structs call deepcopyinit()
// classes call deepcopyinit()

a = <--b
// assigns a mutable reference to a deep copy of `b`
// ensures the reference is a deep copy

a <- ^b
// assigns a mutable copy of `b`
// `b` is consumed and falls out of scope
// structs call moveinit()
// classes call moveinit()

a = ^b
// assigns a mutable reference to `b`
// `b` is consumed and falls out of scope
// structs get treated like reference variables 
// classes call moveinit()

// used as arguments for parameters
fn dostuff(m: Int, n: bor Int, o: mut Int, p: ref Int):
 ...
dostuff(<-a, b, ^c, d)

// `m` passes `a`
// `n` borrows `b`
// `o` takes `c`
// `p` references `d`
```

**Comment by drunkwcodes** [comment]
I also support the use of the `const` keyword. It carries rich historical significance and enhances readability. Let's not wait 30 years like Java to realize its value. Now, Java21 has made a significant update: Hello World no longer requires defining a class! 

![Screenshot_20230622-215711_YouTube](https://github.com/modularml/mojo/assets/36228443/6f61b230-511c-45ca-89dc-66463a128f52)

![Screenshot_20230622-215631_YouTube](https://github.com/modularml/mojo/assets/36228443/34acd517-2952-420c-98c4-5f2ff296c822)


Look at it, Java lags behind Python by 30 years in terms of readability, and even lags behind the C. Before using Python, I used `const` in C++ during my studies, so if I could use `const` in Mojo, this Python family, it would feel very familiar.

**Comment by lattner** [comment]
Hey @gvanrossum :

> On a side note, Python already has something similar to var x: Type [= value] -- you can write x: Type [= value]. Why not use that? It should work in struct as well as in local and function scopes. There's the issue that the syntax corresponding to var x = value has different semantics in Python, but still...

You nailed it.  We need to support that syntax, but we must provide the same semantics Python does or else we break compatibility.  We need some more discussions and iteration, but I think we can probably get away with dropping `let` as a concept.  Even if we do that though, we still need some way to say "this is a lexically scoped variable definition" and the existing `x: Type = expresion()` syntax doesn't do that.  That introduces a new binding, but doesn't go out of scope lexically.  I think `var` as a keyword is a simple and obvious way to do that.

Similarly, we need a way to declare storage in structs and classes.  Classes won't require this, but it will still be a valuable performance optimization for classes if folks choose to use it.  It is required instructs of course.

**Reply by gvanrossum** [reply]
Hm... Inside `fn` and `struct` you can give `x: <type> = <value>` any meaning you want. And in `def`/`class`, why would you need lexical scopes?

**Reply by lattner** [reply]
Philosophically, we want the ability to progressively adopt features a la carte, not have two languages bolted together with a big mode switch named "fn".  Doing so would lead to the potential for people to see def as legacy or deprecated, "the non-mojo thing to do", which isn't what we want.  Such an approach would also hamper progressive learning and teaching, e.g. you'd have to teach a Python programmer about variable scopes and fn all at once, rather than teaching them about 'var' inside of def that they are used to.

IMO, we want def to be a first class feature of Mojo - the right thing to reach for when doing light weight scripting or research.  While it is ok for 'fn' to be a stricter version of 'def', we don't want the core language rules to be different.

**Reply by gryznar** [reply]
Is this approach going to satisfy people who want "static Python"? With control over things like mutability / ownership etc also in global scope?

**Comment by gryznar** [comment]
So maybe a step forward would be make `let` more consistent and unambigous? As someone suggested maybe  `const` (constant - noun/adjective) will be better? This will also match with `var` (variable - noun/adjective).

**Comment by drunkwcodes** [comment]
https://youtu.be/RZhTC33lStQ  22:53 start to discuss the usage of let.

The usage of let is controversial, I also heard `let` in es6 syntax. IMO mojo will confuse a lot of web devs.

But `const` is not that controversial.

**Comment by gryznar** [comment]
I am almost sure, that inspirations to `let` come directly from rust. But in rust `let` is used to declare mutable and immutable variables. Mutable one require also keyword `mut`, so `let` will confuse also rust programmers.

**Reply by martin-braun** [reply]
No, they are inherited from Swift, as the lead dev of Mojo was also the lead dev of Swift. Swift has the very same rules when it comes to `var` and `let`.

**Reply by gryznar** [reply]
Ahh, sure, you are right!

**Comment by martin-braun** [comment]
I'm especially triggered by `let` as its meaning varies across different languages.

- In JavaScript `let` means mutable variable, whereas `const` means immutable variable
- In Swift `let` means immutable variable, whereas `var` means mutable variable
- In Rust `let` means mutable variable when followed by `mut`, otherwise it means immutable variable

I understand. If I was the creator of Swift I'd probably pick the same initially, but even if my goal was the perfect language to rule them all, I should not assume it will be used everywhere. 

In my daily business I have to hop between systems / languages / frameworks very often. I can handle a lot of changes, even the fluctuation of the meaning of `let`, but for the sake of humanity, let's not continue this madness.

How about `const` instead of `let`?

**Comment by david-ragazzi** [comment]
By the way, personally I think it's more intuitive use `const` name instead of `let`. So we would have:

- `var`: declare objects which value is VARIANT (mutable).
- `const`: declare objects which value is CONSTANT (imutable).

More intuitive and clear than that, I think it's impossible.

Furthermore, bellow a scheme (I posted in https://github.com/modularml/mojo/discussions/338#discussioncomment-7178930) where these keywords and the `ref` keyword (to borrow an object) works like a charm:

It consist of:

1. Use the “var” keyword both to declare an object has a variant content (i.e., mutable) and to inform that a function parameter has a variant content (i.e., mutable) and use "const" keyword to say that a object has constant content.
2. Use "ref" as reference to an object, and in the missing of it, the compiler consider the object as owner (in the same way Rust does).

Bellow we have a clue of how these keywords could interate:

```python
const a = "abc"  # a: imutable owned
const b = ref a  # b: imutable reference to "a"

var c = "cde"    # c: mutable owned
var d = ref c    # d: mutable reference to "c"
const e = ref c  # e: imutable reference to "c"
```

Bellow as "const" is the default, we don't need specify it:

```python
def do_something(ref a: String, var ref b: String, c: String, var d: String)
    # a: imutable reference
    # b: mutable reference
    # c: imutable owned
    # d: mutable owned
    ...
```

In the end, we would have only these 3 keywords: `var` and `const` to specify if an object is mutable or not, and `ref` to specify if an object is a reference to another object (in the missing of it, the compiler handle the object as it being the owner of the content).

**Comment by david-ragazzi** [comment]
Several opinions here, but it seems indeed that `const` (instead of `let`) is a common place for everyone and would avoid confusion and resistence.

**Reply by gryznar** [reply]
This is also very well alligned with Rust, where `const` is used for declaring immutable value, which cannot be reassigned (variable declared via `let` in mojo is also not reassignable)

**Comment by codingfun2022** [comment]
``let`` and ``var`` are good choices for the following two reasons.
1. Functional programmers are very familar with ``let``. 
2. ``var`` is short and has the same length with ``let``, which is advantageous for code formatting.

**Comment by gryznar** [comment]
Someone mentioned also `fix` as a unambigous replacement for `let`. This solves problem with different length

**Reply by ivellapillil** [reply]
`fix` is too different from other languages. `fix` also sounds like a verb, while var is a noun. I think size is not that critical. `fix` also does not convey meaning as well as `const`.

**Comment by ksandvik** [comment]
If someone wants to be pedantic, selecting languge keywords should follow a certain pattern:

[keyword - type - grammar construct - longer description]
var  - vartiable entity - noun - variable
const - non-variable entity - noun - constant (where let breaks this symmetry)

similarly
properties - variable properties - adverb - mutable, readable
flow constructs - defines code flow - verb - loop, parallelize()
functions - defines code blocks - verb - run_code()

And there should be no default behavior, every  behavior should be spelled out.

I'm not proposing this directly, more like a possible approach to make keywords and other language structures symmetrical, for example in the case of verbal and written statements. Which leads to inout....

**Reply by helehex** [reply]
!var

**Comment by helehex** [comment]
In theory there is something better than `let`,
but for whatever reason, `let` just feels right to me.
(I mostly stick to one language though)
But I like the other ideas people mention aswell

**Reply by nmsmith** [reply]
`let` reads very poorly when it is not accompanied by an assignment:

```
# Reads nicely
let x = 0
# Reads poorly
let y         # let y what?
# Hypothetical syntax for destructuring
if p is Point(let x, let y):
    do_the_thing()
elif p is Foo(let a):
    do_the_other_thing()
```

In such contexts, `let x` reads like an unfinished sentence.

In my opinion, it's too early to tell which keyword(s) would be most sensible. Mojo needs to have more of its features fleshed out first.

A ideal keyword will make sense in all of the contexts that it appears in.

**Reply by martin-braun** [reply]
@nmsmith It reads poorly, because it is wrong. `let` is a constant, so you have to give it a value.

**Reply by gryznar** [reply]
But Mojo supports variable declaration without assignment, so declaring constant without assigning value should work

**Reply by martin-braun** [reply]
@gryznar 

<img width="769" alt="Screenshot 2023-09-15 at 18 23 09" src="https://github.com/modularml/mojo/assets/24259245/ca32819f-fa1d-489e-b8e5-9a0a63248051">

**Reply by gryznar** [reply]
Ahh, I was always initiating with type, e.g. `let x: Int`. But still I think that `const` suits better :)

**Reply by helehex** [reply]
`let y: Int` is an unfinished statement programmatically.
It presumes a single definition will subsequently be given. So it could be seen as more self-descriptive. But swapping out let for const in most of these cases also works for me, so I think it's just preference. 

My preference is `let` feels nicer to type, looks nicer, and still makes sense in my brain

**Reply by martin-braun** [reply]
@helehex I get where you are coming from. `let x` feels like `let x be it`, which is a passive statement, indicating that you have no active influence over the value. If there wasn't so much confusion from other languages, I wouldn't express my concern.

**Comment by gryznar** [comment]
If team decides to keep `.mojo` files as static only and `.py` dynamic as in Python, (`let` l `const`) could be not needed. All variables could be immutable by default (in global scope also) and allowing them to mutate would require only `var` declaration. Something like in Rust

**Comment by ksandvik** [comment]
Having .py for both python and mojo would be most confusing. Same with different rules for same kind of mojo code and keywords.

**Reply by gryznar** [reply]
You are right, but if Mojo will tend to be strict superset and able to run .py files natively via direct imports (without CPython interoperability), it may be a solution.

The whole point is to determine if default is dynamism or default is strictness

**Comment by lattner** [comment]
FYI, this proposal suggests removing `let` declarations entirely, which would also solve the naming debate :) https://github.com/modularml/mojo/discussions/1456
